### ISSUE TITLE
Autocrafting prioritize other items than in the encoded pattern

### ISSUE THREAD

**Issue by zeromic** [AppliedEnergistics/Applied-Energistics-2#784]
The Autocrafting uses other items as material than in the encoded pattern specified. Those "wrong" items are always prioritized and consumed first in the autocrafting process.

Examples:
"Annihilation Core" encoded with Pure Certus Crystal > Autocrafting uses "vanilla" Nether Quartz.
"Blank Pattern" encoded with Certus Quartz > Autocrafting want to use "Charged Certus Crystal" then "Pure Certus Quartz Crystal" before using "Certus Quartz"
 
Example Screenshot:
![2015-01-20_10 05 34](https://cloud.githubusercontent.com/assets/10611346/5814514/8838f202-a08c-11e4-9a49-e43fae47cbe3.png)
![2015-01-20_10 07 17](https://cloud.githubusercontent.com/assets/10611346/5814515/883fdd74-a08c-11e4-97d8-be1587f5f4b8.png)
Even if there are 50 "Pink Slime" in the Storage, the autocrafting want to use the eight "vanilla" Slime Balls first.

Issue could be ore dictionary related like https://github.com/AppliedEnergistics/Applied-Energistics-2/issues/112

**Comment by yueh** [comment]
That example is an ore dictionary recipe which uses the vanilla slime balls first (part of forge)
And as long as these are available it will use it in the order the recipe defines.
The exact item will only be used if it is not available and needs to be crafted.

It could probably be prioritizing the defined item. But the crafting code is one of the complex parts and I tried for now to stay it if not really needed. Maybe one of the others devs has a better insight.

**Comment by Smashcakes** [comment]
I'm having the same problem with colored wool. I make white wool from cotton, and all my colored wool recipes use white wool in the patterns, but when I autocraft them, they consume the other colored wool first. This wastes the dyes I used to make them in the first place. I crafted 1000 white wool to make sure it was always available, but it still uses the colored wool to make other colored wool.

**Comment by Mephie** [comment]
Same problem here. Autocrafting prefers stained glass to vanilla glass, and vanilla dyes (bonemeal, lapis, rose red, etc) to my Botania Floral powders which are in my paintball recipes.

Running FTB Direwolf20 1.0.4 (MC1.7.10) which is using ae2-rv2-beta-8
--Meph

**Comment by happy2pester** [comment]
I'm having the same issue, FTB Infinity 1.3.1, Rv2-beta-8. If it could prioritize whatever items I specifically put in the recipie, and only after those are exhausted, rely on the oredict, that would be great

**Comment by johnfb** [comment]
I have been having this same issue with silverwood planks. AE wants to use them for making things like crafting tables and pistons. Silverwood should be considered extremely valuable (IMO way more valuable than diamond blocks). (for now I just don't store it in AE, which sucks but whatever). I have thousands of oak logs and it wants to use silverwood!

Perhaps have an option on crafting patterns weather they will use the ore dictionary (say fuzzy) or not? 

This causes some odd issues. It is nice sometimes (say it will use up ingots of ore that are left over from something) but other times it is very frustrating.

**Comment by Darinth** [comment]
Also expressing a lot of support for this. It's actually probably the single most annoying part of my current world. All of the recipes that use wool use all of my colored wool first. All of the recipes that use dye use vanilla dye, rather than the dye that I've selected (Most annoyingly substituting my limited supplies of bonemeal for my limitless supplies of floral dyes from Botania).

**Comment by Balorn** [comment]
Adding another voice of support for this. Running FTB Infinity 1.10.1 (AE2 rv2-stable-10), if a recipe calls for hardened glass (relatively cheap, made from obsidian and lead), the only way I can keep it from autocrafting using fused quartz (made from 4 nether quartz each) is if I remove the fused quartz from the system.

**Comment by yueh** [comment]
I am still not certain, that we can really change or that it is actually feasible (in terms of prefer a different order of oredict entries).

One really important reason is simply the computational effort, just using what is already available is way faster than calculating any potential option. (This is slightly coupled with the next point)
Take enderchests (EnderStorage) for example, there are 4096 different ones. Yes, crafting is limited to the 16 base colors, but you can recolour the wool, there are many mods adding custom dyes for each color etc, and say is it better to recolor a already existing piece of coloured wool to white or use strings to craft a white wool? This will easily add a couple of hundreds or even thousand combinations to explore and it is purely subjective, if `4 string => 1 white wool` is cheaper than `1 blue wool + 1 bonemeal => 1 white wool`.
So AE2 simply tries to fullfil any step with already existing items or if this is not possible, tries the first crafting chain it has available.

The next one is that we are depending on the order the ore dict is providing. So if it tells us that the more expensive item is the first possible option, we will use this first. (It could also be a cheaper option, keep that in mind).
But the oredict is really a design fail like many others in Forge, like it simply has no concept of something like a priority and a couple of other issues (more in terms of implementation details).
So there is simply no way for a mod to say "this is an alternative, but only as last resort".
Also it is completely subjective, if one item is more expensive than another for every player.
Say some player might prefer to replace oak with spruce as first option, but for another it might be cheaper to use greatwood. But for whatever reason, forge is always listing silverwood as first option and we will simply try it. And in terms of a priority, a mod dev might not consider something more expensive than item in mod B. Simply because it is completely unkown to them or having a different balancing than some user wants.

Also how should a player maintain a specific order?
Some entries have thousands of different items and this is the case we have as baseline for it in terms of being usable.
Keep in mind, this is mostly about prefering a (or more) specific item(s) over others. Not the case use oredict at all.

**Comment by Balorn** [comment]
At the least, I'd think it should try for an exact match before asking oredict for a substitute. I understand setting priorities would be rather involved, but I'd think a check with oredict off first should be possible.

I know various machines/filters exist that have options to use oredict or not, so it seems like there should be an option to enable/disable oredict for crafting (ideally, on a per-recipe basis, but failing that, a global option, or maybe a per-Interface option).

**Comment by yueh** [comment]
Completely disabling oredict is something completely different, and there is another issue + PR for it.

**Comment by happy2pester** [comment]
You could potentially do something like disable oredict per recipie, i know that's been discussed before

**Comment by Darinth** [comment]
I don't think anybody in this thread is interested in seeing each individual recipe or user capable of customizing the priority list of items that are consumed by a recipe. We're really just interested in A: having the autocrafting system default to trying the exact recipe programmed first, and B: having the ability to disable the ore dictionary all-together (either on a per-recipe or per-interface basis), to ensure more valuable resources aren't used.

**Comment by Blue64** [comment]
It's a simple If/Else Array really.

Get the total count of the various items selected, which I assume is already done.

If the exact items exist in the network, request them, otherwise, revert to standard behavior.

Plenty of mods do have optional OreDict toggle, and they do it on a per terminal basis ([Refined Relocation] [Extra Utilities] [Random Things 2] to name a few).

The masses are just wondering why they need to outsource their OreDict toggle when this is supposed to be a High-End Ultra-Tech Mod, complete with Vibration chambers, an Item Network, complete automation, and various other functions.

How exactly can one have complete all in one automation if they need to go elsewhere for their sorting needs (which totally breaks the whole "item network" idea By The Way)?

**Comment by thatsIch** [comment]
I think the idea is to try to perform crafting with the exact ingredients first and use oredict as last resort.

This will probably require a rework of the recipes system to make it work because currently a oredictionaried recipe discards the knowledge of the original and will use all synonyms of it to perform the recipe.

**Comment by Blue64** [comment]
I was actually thinking you could just add an if/else check just before it heads for the oreDict to check the network for exact count first, it'd be much simpler that way instead of reworking the entire setup, just gotta make sure it won't delete items on ya or ignore em (for example item names that contain a "." sometimes are problematic), shouldn't be too difficult. I'm not suggesting it needs a full overhaul, just throw in an extra check to compare the amount needed of the items set to the network's available inventory. Probably want to do it on a per iteration basis instead of the full loop (eg crafting 50 Rose Red favoring the small flower until it runs out, then moving on to the double flower after empty because the result is the same, or vice versa, based on which was set for the recipe by the player, run once per craft ping to the network for materials instead of checking when the craft amount is set)

**Comment by thatsIch** [comment]
It is not about difficult, it is about people loosing all their recipes. We have to keep compability as high as possible between versions so that people do not loose their progress. Recipes which were created 1 month ago without knowledge with which item they were created will still not know 1 month later what they were.

**Comment by Darinth** [comment]
I'm confused by that statement. I know patterns maintain, at least in some format, their original item list. If I stick a pattern that uses the ore dictionary back into the pattern terminal it properly populates the pattern terminal with the original recipe I used. Am I misunderstanding what you're saying thatsIch?

**Comment by Blue64** [comment]
I too am confused, I'm seeing my Red Dye substituted for the item I put in, complete with a text readout of the exact items used. Even if they need to cycle their Crafting Patterns once through the terminal just to make sure, I honestly don't think people would mind considering the flexibility it would add to the infrastructure.

Perhaps you want to use charcoal for torches, instead of coal, but the server admin is a jerk and labeled them under the same OreDict via MineTweaker (which I've actually done, looks so much better in NEI that way) and also finds it hysterical that you are breezing through your Coal supplies which you're trying to (painstakingly) get a decent sized buffer of.

As it currently stands, you'd need to outsource this task, which is just nonsensical, an auto-crafting CPU that can't handle a simple variable grab from the pattern it's already grabbing from to explicitly attempt the pattern first, before resorting to OreDict, nonsensical I tells ya

Simply put, there's a wide variety of reasons why people want it to be capable of understanding that people want it tried a specific way first before they resort to "don't care, gimme!". The simulation GUI in place offers a simple interface for this option, it can even simply state if the simulation was able to follow explicit or if it had to resort to OreDict.

I know I would absolutely love this feature, I don't see how it's game breaking since it can easily look at the names in the encoded pattern (Meta/NBT if I'm not mistaken) with a simple array to split up the items accordingly. I may suck at arrays, but even I can see this working.
